### ISSUE TITLE
add 'backwards' mode for blurring

### ISSUE THREAD

**Issue by richlv** [mapillary/mapillary_issues#2161]
the 'clear all and submit' button now advances to the next image in the sequence. it would be nice to allow moving backwards, too - sometimes i start reviewing sequences from the middle or so.
this could be either a control that could be flipped and the buttons would change their operation, or the buttons could have different operation performed on the left/right side

**Comment by gyllen** [comment]
@angelday  I think what we probably want is a set of keyboard shortcuts for `power users` such as @richlv

**Comment by angelday** [comment]
I'd like to support such use case (submit + forward/back) for more advanced users and simplify the UI for the common use case. Today we have a mix from both worlds.

**Comment by gyllen** [comment]
Yeah I would say simplicity in the buttons and then a bunch of nice keyboard short cuts.

@richlv Any suggestions other than backwards and blur?

**Comment by richlv** [comment]
whether it is a switch for fwd/back or two sides of the button, that doesn't seem to matter that much. unless the keyboard shortcut is very easy to use, buttons would be more preferred.
jumping to the next image with blur would be great, but that's a separate request :)

**Comment by angelday** [comment]
I like the idea.

**Comment by de-vries** [comment]
I want to share my ideas about the controls I need for an optimal work-flow for the use-case of reviewing all photos in a sequence.

_In almost all photos  I need one of this three controls:_
**move to next photo**
When no modifications are needed I need to go to the next photo in the sequence. Now this can only be done with the arrows in the top. The up/down arrow keys often take you to another sequence, or skip a photo, what makes it impossible to track your progress. My solution would be to bind the left/right arrow keys to previous/next in sequence. With the positive side-effect that this can be implemented site-wide and would eliminate most need for a 'lock to sequence' feature.

**clear all submit and move**
Exactly how this function works now, but with a hot key `c` would be a logic choice. `Shift-c` then could be used to `clear all submit and move back`.

**submit and move**
After manually adjusting blurs I want to submit and move to the next photo. This shouldn't be the default action as it might confuse people, but for power users it makes it a lot faster.
I see two options for key bindings here
The first would bring a nice symmetry to the `clear all` key-bindings I proposed:
 - `s` submit and move
 - `Shift-s` submit and move back

The second just adds a modifier to the navigation keys:
- `Shift-right arrow` submit and move to next in sequence
- `Shift-left arrow` submit and move to previous in sequence
- `Shift-up arrow` submit and move to next photo
- `Shift-down arrow` submit and move to previous photo

**Comment by GITNE** [comment]
Yeah, generally speaking the blurring editor needs shortcut support and arrow key navigation. I am all for @de-vries suggestions as long as they are transparent to the user.

**Comment by gyllen** [comment]
@de-vries I like it

**Comment by GITNE** [comment]
Hmm, apparently the blur editor does have arrow key navigation, however this refers to navigation in space, not in the sequence. I think having both would be great. Perhaps ```Page-Up```/```Page-Down``` would be nice for navigating the sequence?

**Comment by GITNE** [comment]
@gyllen Anyhow, the new blur editor has become much more ergonomic and agile than in the previous version. Latency has reduced dramatically too. :+1: for that! :smiley: There are still some quirks (like keyboard support) but you're on a good path. Keep it up! :smile:

**Comment by gyllen** [comment]
@GITNE Hey, thanks! Nice to hear. We have been a small team a long time but finally managed to hire some new (super skilled) people to help us keep up on all areas.

So keep on filing issues and keep pushing us. Your feedback are always appreciated!

**Comment by richlv** [comment]
@GITNE, there are rumours that alt+up/down might move in sequence, but that doesn't seem to be documented, so not that sure about it :)

**Comment by gyllen** [comment]
@GITNE Not to my knowledge. We are right now discussing short cuts though. Do you have any suggestions?

**Comment by richlv** [comment]
in general, i'd prefer buttons for most things except maybe fwd/back/clear-all/submit. and a toggle (a button) for "lock to sequence". all other things would be better as ui buttons.
so arrows for fwd/back, maybe 'enter' for submit and smth for clear all - with the last two being less important.

if you really, really think about creating a power user tool, do it as a josm plugin. that will be easier in the end in so many ways.

**Comment by GITNE** [comment]
> @GITNE, there are rumors that ```Alt+Up```/```Down``` might move in sequence, but that doesn't seem to be documented, so not that sure about it :smiley:

Indeed, this key combination does traverse the sequence. :laughing: Unfortunately, this is not transparent to the user. Personally, I would prefer ```Page‑Up```/```Page‑Down``` because it is not a key combo, easier and faster to grab than a key combo, and the blur editor is not designed to have scrolling pages anyway. So, this makes a natural choice IMHO.

Besides, I have found a bug while using this key combo. The key combo only works when focus is on the ```div class="ButtonContainer flex flex-column justify-between items-center height-100 absolute top-0 right-0 pt2"``` element. If the focus is on the photo container it does not work, at least in Firefox. I do not know about other browsers.

> @GITNE Not to my knowledge. We are right now discussing short cuts though. Do you have any suggestions?

Initially, this issue was a request for a ```Submit + <``` button or switch to set the default sequence traversal direction. However, I think the discussion has evolved into a discussion on general usability of the blur editor, which I think is good. With regards to shortcuts and keyboard support, then as I said before, IMHO ```Page‑Up```/```Page‑Down``` would be a more natural choice for traversing the sequence. Generally speaking, I would try to refrain from key combos for the ease of use because pressing one button is always easier or simpler than grabbing a key combo. Of course, single keys may have a higher potential to collide with exiting browser or OS functionality. So, you would have to test it on major browsers before deployment.

> if you really, really think about creating a power user tool, do it as a josm plugin. that will be easier in the end in so many ways.

Yep, evolving the Mapillary plug-in for JOSM into a advanced/power user tool seems like a good idea to me too. However, this applies only to me because I am familiar with JOSM and have done quite an amount of OSM editing in it. Referring other professional people to JOSM from different professional backgrounds, like say photography, might not be the best thing to do because they are usually not interested in mapping. Of course, you do not have to learn mapping in JOSM in order to use the Mapillary plug-in, yet the learning curve in JOSM is quite steep or may be quite daunting when confronted with for the first time.

Having said that, I think one should deal with one tool at a time. So, adding shortcut and keyboard support to the web based blur editor while making it more user friendly at the same time is a good thing. I do not see any reason why the web based blur editor should not be made very helpful and usable to novice **and** advanced users. Sure, some features may be better left to implement into the Mapillary plug-in for JOSM but this is a topic of its own. Besides, no tool is perfect. :wink:

**Comment by gyllen** [comment]
@GITNE I agree with everything above. We have users from many different categories and having them install JOSM can be.... quite a challenge. I love JOSM but its the mother of all power tools.

We should focus on making web experience great but also put more focus on clearly document the main three things in the mapillary platform `MapillaryJS`, `Vector Tiles` and `API`. This way we can encourage people to help us build the JOSM plugin.

Also we are still wrapping our head around shortcuts, suggestion will be coming soon.

**Comment by smsm1** [comment]
I'd really like to have keyboard shortcuts to save my hands from RSI with repetitive moves of the trackball. Also note that the compact Apple keyboards (which I use) required require `fn`+`down arrow`/`up arrow` to get page down/up, so letter keys like in vim command mode would be preferred.

**Comment by smsm1** [comment]
Oh there is also #360 that I started ages ago about keyboard shortcuts. The new uploader would be useful to have keyboard shortcuts too, particularly for deleting image, and also selecting the next image in the sequence after a delete.

On the choosing previous or next photo to select after deleting, you could follow the Apple Mail approach where it remembers the direction you have be going over the past few messages, and then selected the next message after deletion based on that, so that you don't end up on a message you've just looked at.

**Comment by richlv** [comment]
@GITNE, at least on my systems, blur editor often gets vertical scrolling, thus up/down is annoying sometimes already.
regarding the JOSM usage, i didn't suggest that the web interface shouldn't be made better - just that investing a lot of time in it to recreate what josm has to support advanced users might not be the best course right now :)

@gyllen, note that i only suggested leveraging josm for features that are not trivial to do in the web interface. as for installing it, i've suggested using https://josm.openstreetmap.de/download/josm.jnlp to some novice mappers - that's not too hard to do ;)

another thing why a josm plugin would be better for power users (and please note again that's the only category of users i'm suggesting would be subjected to that :) ) is performance.

latest version of firefox, a bix with 16gig of ram and a fast ssd - the web interface is slow and the computer sounds like an airplane that's about to lift from my desk.
trying to access the web interface in a vm is painfully slow, pretty much unusable. i can run josm in that vm perfectly (and it's java, take that).

i don't want to whine (too much) or belittle the developers - but mapillary has had an online blur editor for what, a few years now ? and it is still pain to use... reusing what josm has might have brought us more of a power tool for blur editing.

**Comment by GITNE** [comment]
> I'd really like to have keyboard shortcuts to save my hands from RSI with 
> repetitive moves of the trackball. Also note that the compact Apple keyboards 
> (which I use) required require |fn|+|down arrow|/|up arrow| to get page down/up, 
> so letter keys like in vim command mode would be preferred.

@smsm1 If you are using a non-standard keyboard then it is **your** **personal** **choice**. You can not expect others to adhere to or to support non-standard hardware. So, you do not have an argument here. However, what everybody can and should expect from any vendor is to comply with standards.
AFAIK, no web browser currently exposes any JavaScript API to query for attached keyboards. In Apple's case it may suffice to query for the Safari agent. But then again, third party web browser users (like e.g. Firefox) might run into the same issue on non-standard Apple systems.

> On the choosing previous or next photo to select after deleting, you could follow the Apple Mail approach where it remembers the direction you have be going over the past few messages, and then selected the next message after deletion based on that, so that you don't end up on a message you've just looked at.

This sounds like a reasonable idea indeed. The blur editor could start without either ```Submit + >``` or ```Submit + <``` button but a```Submit```‑only button on the first photo. Then the user would have to advance to the next photo manually while also choosing direction. After that the assumed direction could change any time the user would switch direction manually.

> @GITNE, at least on my systems, blur editor often gets vertical scrolling, thus up/down is annoying sometimes already.

Yeah, I was not sure if no scroll bars could be assumed in all web browsers but it should definitely be possible to design the page in a way where scroll bars should never appear (unless they are *always* displayed by the browser). Anyway, it should be possible to consume any ```Page‑Up```/```‑Down``` key events without letting the event trickle up to the browser for default processing.

> as for installing it, i've suggested using https://josm.openstreetmap.de/download/josm.jnlp to some novice mappers - that's not too hard to do

This makes two huge assumptions on novice users: 1. the user has properly installed a JRE and 2. a JNLP plug‑in for their browser.

> latest version of firefox, a bix with 16gig of ram and a fast ssd - the web interface is slow and the computer sounds like an airplane that's about to lift from my desk.
> trying to access the web interface in a vm is painfully slow, pretty much unusable. i can run josm in that vm perfectly (and it's java, take that).

@richlv Well, this is due to **your** **personal** **setup**. If you want to run anything in a VM then it is your choice. Unless a software is purposely built to run in a VM, no developer can take into account that his/her application is going to be run in a VM. Besides, if you are having performance problems running a browser in a para‑virtualized VM then you are definitely doing something wrong. Any performance penalty incurred by a para‑virtualized VM should be close to zero. Your performance problems are most probably due to your virtualization solution lacking support for hardware accelerated graphics (which modern browsers use lot of), not due to Mapillary's web page.

> i don't want to whine (too much) or belittle the developers - but mapillary has had an online blur editor for what, a few years now ? and it is still pain to use... reusing what josm has might have brought us more of a power tool for blur editing.

I would not go as far to say that especially the new blur editor is a pain to use, even for power users. It is definitely a lot faster than in the previous version. However, there are still some quirks like missing keyboard support, inability to draw blur rectangles in any direction, unnecessary 3D panning, and unnecessary transitions.

**Comment by richlv** [comment]
@gitne, let me clraify. mapillary image viewer in general is slow even on the bare, non-virtualised hw. i was just adding a comparison that josm runs perfectly fine even in a limited vm (a different setup) while mapillary website becomes completely unusable there.

**Comment by GITNE** [comment]
@richlv Thank you for the clarification. However, your comparison is flawed. It is like comparing apples and oranges.

**Comment by smsm1** [comment]
@GITNE I'm using a standard keyboard that comes with pretty much all Macs, and the mention of it was simply to have the option of single key presses taking account of the fact there is no page up/down key on these and other keyboards. Taking a quick look online there are several other brands of keyboard which don't have a dedicated page up and down key, requiring you to also use the fn modifier key to be able to get the functionality.

**Comment by smsm1** [comment]
Also the fn+page up/down sends the normal page up/down command, thus no keyboard specific info is required. It is the keyboard itself that converts the extra keyboard press to the standard command.

**Comment by iav** [comment]
>  fn+page up/down sends the normal page up/down command

Better not use two-handed combinations. fn+arrows not just 2-keys, but 2 hand.

**Comment by GITNE** [comment]
> I'm using a standard keyboard that comes with pretty much all Macs, and the mention of it was simply to have the option of single key presses taking account of the fact there is no page up/down key on these and other keyboards.

@smsm1 Obviously you have a problem understanding what *the* or *a* standard is. Just because your non‑standard keyboard comes with every Mac does not mean it is a standard keyboard. Again, it has been **your** **personal** **choice** to choose a non-standard keyboard from Apple, so your complain is inadmissible as an argument. Trying to avoid key combinations for ergonomic reasons where possible is a good design goal. However, the only keyboard developers can safely assume to either require or not require key combinations are the standard keyboards (if they are attached is a different story). Standard keyboards in fact do not require a key combination for ```Page‑Up```/```‑Down```. So, my well thought‑out suggestion still stands, despite your invalid interjection about key combinations on non‑standard keyboards. If developers would not be guided by standards in their design decisions then the next person with another non-standard keyboard would come along and start complaining that he/she has to press some other key combination to get the same standard keyboard behavior. In the end you will not be able to find a key stroke or key combination that is going to please everybody. The **standard** is what everybody has silently (unless you are a member of the standards body) agreed on. Live with it or buy a new standard keyboard. Or, propose a new standard *if* you have convincing arguments why your proposition should become a standard (warning: may lead to a huge loss of investment of time and money, including unfulfilled lifelong dreams and despair). :laughing:

> Taking a quick look online there are several other brands of keyboard which don't have a dedicated page up and down key, requiring you to also use the fn modifier key to be able to get the functionality.

A myriad of non‑standard keyboards exist, especially if you look at notebook keyboards. Some of them have the strangest key combinations to produce standard keyboard key strokes. But, you can not possibly account for every single one of them. It would be stupid to try to do so. Again, the only keyboards a developer can safely design for are the standard keyboards.

**Comment by richlv** [comment]
@gitne, this is getting a bit too aggressive. i'm not a fan of apple by far, but the point @smsm1 raised was quite valid, and a bit more civil argument would be appreciated, thanks :)

**Comment by GITNE** [comment]
@gyllen I wish this — which is keyboard accessibility in general — would finally get resolved. :crossed_fingers: #2269 and #2741 basically all call for similar issues to address.

Reviewing the discussion here after such a long time makes me think that @de-vries’ [keyboard shoutcut suggestions](#issuecomment-238650265) seem quite reasonable. Well, maybe except for `Shift`+`↓` which is already reserved by space navigation (turn around). So, perhaps `Shift`+`Alt`+`↑`/`↓` may be an alternative? This should not collide with `Ctrl`+`Alt`+`↑`/`↓` on some X-Window systems either.

**Comment by gyllen** [comment]
@angelday Some good feedback ^^^

**Comment by angelday** [comment]
Keeping it as blur editor improvements. (Note: blur editor is currently offline.)

_mapillary_issues is no longer an active repository._

_Let’s continue discussing features in our forum! We will import this issue as topic to the Mapillary Forum in the coming weeks. We’ll notify here once it’s been done. See you there!_
